src/databricks/labs/ucx/framework/owners.py

```diff
@@ -169,8 +171,21 @@ def get_workspace_administrator(self) -> str:
 class Ownership(ABC, Generic[Record]):
     """Determine an owner for a given type of object."""
 
+    _ownerships: set[Ownership] = set()
+
+    @classmethod
+    def for_record(cls, record: Any) -> Ownership[Record]:
+        for ownership in cls._ownerships:
+            if ownership.is_applicable_to(record):
+                return ownership
+        raise ValueError(f"Ownership not implemented or not registered for {type(record).__name__}")
+
     def __init__(self, administrator_locator: AdministratorLocator) -> None:
         self._administrator_locator = administrator_locator
+        self._ownerships.add(self)
+
+    @abstractmethod
+    def is_applicable_to(self, record: Any) -> bool: ...
```

Comment:
Sorry, I missed [this comment](https://github.com/databrickslabs/ucx/pull/3072#discussion_r1824480883) indicating that we want to allow instances to support multiple types.

That said, I don't fully follow the reasoning: type-based lookups can also support multiple instance types (`A | B`) and encourage a single lookup rather than the (current) situation that encourages callers to implicitly linear probe the registry within a loop. I think I'm missing something here, and hope that @nfx can help me understand what he has in mind.